bindings/python/classad3/ClassAd.py

```diff
@@ -1 +1,266 @@
+        :return: A :class:`list` of :class:`str`\\s of attribute names
+            referenced in *expr* but not defined in this ClassAd.
+        '''
+        return expr.externalRefs(self)
+
+
+    def internalRefs(self, expr : 'Expression') -> List[str]:
+        '''
+        Equivalent to ``expr.internalRefs(self)``.
+
+        :param Expression expr: The expression to consider in this context.
+        :return: A :class:`list` of :class:`str`\\s of attribute names
+            referenced in *expr* and defined in this ClassAd.
+        '''
+        return expr.internalRefs(self)
+
+
+    def matches(self, ad : 'ClassAd') -> bool:
+        '''
+        Returns ``True`` if the ``requirements`` expression in
+        *ad* evaluates to ``True`` in the context of this ad, or
+        ``False`` otherwise.
+
+        :param ClassAd ad: The ad to match.
+        :return: If *ad* matched this ad.
+        '''
+        try:
+            r = self["requirements"]
+        except KeyError:
+            return False
+
+        # FIXME: Not sure of the semantics here, see below.
+        if not isinstance(r, Expression):
+            return bool(r)
+
+        # FIXME: check the semantics here.  The docstring implies
+        # that this shold be `is`, but in practice matches() may have
+        # always returned `True` if the requirements expression was
+        # considered true (for example, a nonzero integer), as well.
+        return bool(r.evaluate(my=self))
+
+
+    def symmetricMatch(self, ad : 'ClassAd') -> bool:
+        '''
+        Equivalent to ``self.matches(ad) and ad.matches(self)``.
+
+        :param ClassAd ad: The ad to match against.
+        :return: Whether the this ad matched that ad and vice-versa.
+        '''
+        return self.matches(ad) and ad.matches(self)
+
+
+    def asNewClassAd(self) -> str:
+        '''
+        Serialize this ClassAd into the "new" ClassAd format.
+
+        :return:  This ClassAd serialized in the "new" ClassAd format.
+        '''
+        if len(self.data) == 0:
+            return '[ ]'
+
+        newClassAd = '[ '
+        # It might actually be faster to write this as `"\n".join([...])`
+        # where [...] is a list comprehension because that will only
+        # allocate the final string once.
+        for key, value in self.data.items():
+            value_str = _valueAsNewClassAd(value)
+            newClassAd += (f"{key} = {value_str}; ")
+        newClassAd = newClassAd[0:-2] + " ]"
+        return newClassAd
+
+
+    #
+    # Restrict the keys and values appropriately.
+    #
+    def __setitem__(self, key, value):
+        # Let's be paranoid to simplify the C++ bindings.
+        if not isinstance(key, str):
+            raise KeyError("Key must be str.")
+
+        # Permit only "old" ClassAd attribute names for now.
+        match = re.match('^[a-zA-Z_][a-zA-Z_0-9]*$', key)
+        if match is None:
+            raise KeyError(f"Key '{key}' invalid.")
```

Comment:
Might be good to [compile](https://docs.python.org/3/library/re.html#re.compile) this regex if it's potentially being used a lot (I think `__setitem__` qualifies). Feels like there should be an existing function since these restrictions are the same for Python object names? Aha, maybe [str.isidentifier()](https://docs.python.org/3/library/stdtypes.html#str.isidentifier):
```suggestion
        # Permit only "old" ClassAd attribute names for now.
        if not key.isidentifier():
            raise KeyError(f"Key '{key}' invalid.")
```

Reply:
Hmm, apparently Python accepts non-ASCII characters:
```python
>>> "Øops".isidentifier()
True
```

Reply:
Agreed that it would be better to the compile the regex.  For now, I'll just note that in a comment.